### ISSUE TITLE
A specific network for MNIST is hard to train with TFJS

### ISSUE THREAD

**Issue by photopea** [tensorflow/tfjs#2664]
There exists a specific network for MNIST, that is trained and executed with ConvNetJS library, shown here: https://cs.stanford.edu/people/karpathy/convnetjs/demo/mnist.html  . It has four layers.

I tried to rebuild it with TFJS, and make all parameters identical to ConvNetJS version (including parameters of Adadelta learning, batch size, etc., I trained both without l2 decay):

    model.add(tf.layers.conv2d      ({inputShape: [24,24, 1], kernelSize: 5, filters: 8, activation: 'relu', padding:"same"}));
    model.add(tf.layers.maxPooling2d({inputShape: [24,24, 8], poolSize  : 2, strides: 2}));
    model.add(tf.layers.conv2d      ({inputShape: [12,12, 8], kernelSize: 5, filters:16, activation: 'relu', padding:"same"}));
    model.add(tf.layers.maxPooling2d({inputShape: [12,12,16], poolSize  : 3, strides: 3}));
    model.add(tf.layers.flatten({}));
    model.add(tf.layers.dense({units: 10, activation: 'softmax'}));
    model.compile({optimizer:'adadelta', loss: 'meanSquaredError'});
    ...
    model.compile({optimizer:tf.train.adadelta(1, 0.95, 1e-6), loss: 'meanSquaredError'});

I trained it with just one iteration over all 60 000 MNIST samples. The ConvNetJS version makes about 140 errors after one training iteration, while TFJS version makes about 220 errors after one iteration, and about 150 errors after four iterations. Even after trying it several times, the training quality of TFJS did not get even close to CNJS (TFJS had 270 - 190 errors, while CNJS had 150-110 errors).

Am I doing something wrong? At the moment, it looks like CNJS will give a network with a sufficient quality faster than TFJS, even though one training iteration takes longer (as it needs a smaller number of iterations).

**Comment by rthadur** [comment]
@photopea we can not comment on CNST but regarding can you please provide these details 
1) TFJS version and node version.
2) if possible can you try to run in [codepen](https://codepen.io/pen/)  and share it with us.

**Comment by photopea** [comment]
I am testing it in a browser, in the latest Chrome, with the latest TFJS https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@1.0.0/dist/tf.min.js 

I think Javascript and WebGL works the same everywhere (with varying speed, but the same computatoin mechanism and results). I am just curious, if TFJS can train a specific network (four layers + softmax, described above) to get under 150 errors on MNIST dataset, with only one iteration over the training data, as other engines can.

**Comment by nsthorat** [comment]
Hi @photopea, can you retry the experiment by setting `tf.setBackend('cpu')` before running the TensorFlow.js model? This will help us understand if it's 1) initialization of weights or 2) numeric stability of the GPU.

**Comment by photopea** [comment]
Hi, I just tried, the results are similar - the network makes 231 errors after training. Is TFJS using any learning rate? I think at Adadelta, it should be one.

Also, the training with TFJS on the CPU took 45 minutes, while CNJS (also on the CPU) takes 4 minutes.
